### PR TITLE
Fix oracledb_session SQLCL and CSV output parsing

### PR DESCRIPTION
**Pull request by gsreynolds**
## Description

The `oracledb_session` resource is non-functional against the following: `Oracle Standard Edition One (11.2.0.4.v18)` on RDS using `sqlcl-18.3.0.259.2029`

* sqlcl_bin option isn't used if provided
* silent options for sqlcl and sqlplus binaries not passed, need to remove command prompt, command echoing and connection banners to avoid breaking CSV parsing
* Fix 'no implicit conversion of Regexp into String' with string `.delete`
* Leading blank line in output breaks CSV parsing

Potentially related to #3594

The CSV output is preferred but appears not to be covered with the unit test mocking (test/unit/mock/cmd/oracle-result is the HTML format)

## InSpec Version
3.0.52 (last change to `oracledb_session` resource was v2.2.35 in https://github.com/inspec/inspec/pull/3170)

## Replication Case
```ruby
control 'test' do
  impact 1.0
  title 'Test'
  desc 'Test oracledb_session'
  sql = oracledb_session(user: '<USER>', password: '<PASS>', host: '<HOST>', port: '1521', service: 'ORCL', sqlcl_bin: '<PATH TO SQLCL>')
  describe sql.query('select username, profile from dba_users;').column('profile') do
    it { should include 'DEFAULT' }
  end      

  describe sql.query('SELECT NAME AS VALUE FROM v$database;').row(0).column('value') do
    its('value') { should eq 'ORCL' }
  end
end
```

## Output
Before:
```
Profile: InSpec Profile (oracledb)
Version: 0.1.0
Target:  local://

  ×  test: TEST (2 failed)
     ×  [] should include "DEFAULT"
     expected [] to include "DEFAULT"
     ×  SQL Column value should eq "ORCL"

     expected: "ORCL"
          got: ""

     (compared using ==)
```
After:
```
Profile: InSpec Profile (oracledb)
Version: 0.1.0
Target:  local://

  ✔  test: TEST
     ✔  ["RDSADMIN", "RDSADMIN", "DEFAULT", "RDSADMIN", "DEFAULT", "DEFAULT", "DEFAULT", "RDSADMIN", "DEFAULT"] should include "DEFAULT"
     ✔  SQL Column value should eq "ORCL"

```

### CODE REVIEW

**Comment by btm** [comment]
@gsreynolds can you look at the travis and appveyor failures on this?

**Comment by gsreynolds** [comment]
Hi @btm, from what I can see the failures are unrelated to the PR. 

All the Travis failures are:
```
$ rvm @global do gem install bundler
bundler's executable "bundle" conflicts with /home/travis/.rvm/rubies/ruby-2.3.6/bin/bundle
Overwrite the executable? [yN]  
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

The two Appveyor fails are apparently also unrelated:
```
inspec json::inspec json with a inheritance profile#test_0002_populates code from child profiles:
NoMethodError: undefined method `[]' for nil:NilClass
    C:/projects/inspec/test/functional/inspec_json_profile_test.rb:145:in `block (3 levels) in <top (required)>'
```

I've rebased on master and force pushed to see if that clears the failures.

**Comment by gsreynolds** [comment]
Hi @btm, I fixed an oversight on my part in `test/helper.rb` which has fixed the unit tests and Travis now passes.

However, Appveyor is still suffering from random failures, Appveyor passed once but when I tided my git history and pushed again, it has now failed again :(

File: lib/resources/oracledb_session.rb
Comment (clintoncwolfe) [inline lib/resources/oracledb_session.rb:44]: The comment mentions something called sqlci, the option is named sqlcl, and the actual value is either 'sql' or 'sqlplus'.  Could you clarify?  Assume the maintainer of this code is not familiar with Oracle.

Comment (clintoncwolfe) [inline lib/resources/oracledb_session.rb:44]: This appears to introduce a new resource parameter, `sqlcl_bin` but does not document that in the documentation.

File: test/helper.rb
Comment (clintoncwolfe) [inline test/helper.rb:476]: Why does the hash change?  I don't see a matching change to a command mock file...

File: lib/resources/oracledb_session.rb
Comment (clintoncwolfe) [inline lib/resources/oracledb_session.rb:71]: A set of comments picking these apart - or links to online Oracle docs - would help the maintainability of this code. A lot happens in different ways, and I don't see the reasoning.

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Thanks for the contribution! However, since this adds a new resource parameter, we need to document that.  Also, clarifying comments would really help; the InSpec team does not work with Oracle databases very much, so anything you can do to explain the code to people unfamiliar with it would be great.

Finally, and I understand this is likely out of scope, I would feel a lot more comfortable if we had real integration tests. While a unit test based on canned output is a good start, I would feel a lot better about this if we had a set of rake tasks to setup, test more thoroughly, and then teardown some RDS instances; something like #3684.

File: lib/resources/oracledb_session.rb
Comment (gsreynolds) [reply lib/resources/oracledb_session.rb:44]: `sqlcl_bin` was an existing resource parameter that is documented but was previously unused in the code.
https://github.com/inspec/inspec/blob/8cdd7cfe0309528e2c866e1c311ee854e012e94a/docs/resources/oracledb_session.md.erb#L34

Comment (gsreynolds) [reply lib/resources/oracledb_session.rb:44]: The comment is incorrect, i've corrected it to `sqlcl`

File: test/helper.rb
Comment (gsreynolds) [reply test/helper.rb:476]: The hash changed as the `-S` silent options were added in the commands (I'm not sure why there isn't a mock file)

File: lib/resources/oracledb_session.rb
Comment (gsreynolds) [reply lib/resources/oracledb_session.rb:71]: This logic was added in the previous PR that touched this resource: https://github.com/inspec/inspec/pull/3170. The way the logic is constructed appears to do:
1. User & password authentication
2. User & password authentication plus assuming a database role
3. User & password authentication plus assuming a database role executing the binary as root

File: lib/resources/oracledb_session.rb
Comment (clintoncwolfe) [reply lib/resources/oracledb_session.rb:44]: Ahh, thanks!

File: lib/resources/oracledb_session.rb
Comment (clintoncwolfe) [reply lib/resources/oracledb_session.rb:71]: Yes, and I found it confusing then, as well.  What I was asking for here is comments *in code* explaining what is going on, or a URL *in a code comment* where a maintainer with little knowledge of Oracle can go to understand what is happening, and why it happens three different ways.

**Review by miah — CHANGES_REQUESTED** [review]
Thanks for the contribution. This looks solid, could you address the changes requested by @clintoncwolfe after that this has my approval.

**Comment by clintoncwolfe** [comment]
(boilerplate message)
Hi! This PR has been untouched for more than two weeks, in the "Changed Requested" state. Please either make the requested changes, or continue the discussion in comments.

**Review by miah — APPROVED** [review]
I think this fixes a few issues we're currently dealing with, so lets get this pulled through. @clintoncwolfe can you give this another look over? We can add the comment about the login stuff afterwards (I'll do it).

**Review by clintoncwolfe — APPROVED** [review]
Thanks!

**Comment by miah** [comment]
I'm leaning towards getting #4654 merged and closing this one out. The only additional that this one does is with sqlcl, but there are no tests. I have sqlcl locally,  so i can start to build that out. But I want to get sqlplus into a zone of stability first =)

**Pull request closed without merging**
